## Data Analysis with Pandas on IMDB Movie Dataset

Objective: To find some interesting insights into a few movies released between 1916 and 2016, using Python pandas.

In [178]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [179]:
# Import csv dataset
movies = pd.read_csv('Movies.csv')
movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4     Color     Andrew Stanton                   462.0     132.0   
...     ...                ...                     ...       ...   
3848  Color      Shane Carruth                   143.0      77.0   
3849  Color   Neill Dela Llana                    35.0      80.0   
3850  Color   Robert Rodriguez                    56.0      81.0   
3851  Color       Edward Burns                    14.0      95.0   
3852  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
0                         0.0                   855.0    Joel David Moore   
1                       563.0                  1000.0       Orlando Bloom   
2                         0.0                   161.0        Rory Kinnear   
3                     22000.0                 23000.0      Christian Bale   
4                       475.0                   530.0     Samantha Morton   
...                       ...                     ...                 ...   
3848                    291.0                     8.0      David Sullivan   
3849                      0.0                     0.0     Edgar Tancangco   
3850                      0.0                     6.0     Peter Marquardt   
3851                      0.0                   133.0  Caitlin FitzGerald   
3852                     16.0                    16.0    Brian Herzlinger   

      actor_1_facebook_likes        gross  \
0                     1000.0  760505847.0   
1                    40000.0  309404152.0   
2                    11000.0  200074175.0   
3                    27000.0  448130642.0   
4                      640.0   73058679.0   
...                      ...          ...   
3848                   291.0     424760.0   
3849                     0.0      70071.0   
3850                   121.0    2040920.0   
3851                   296.0       4584.0   
3852                    86.0      85222.0   

                                   genres  ... num_user_for_reviews language  \
0         Action|Adventure|Fantasy|Sci-Fi  ...               3054.0  English   
1                Action|Adventure|Fantasy  ...               1238.0  English   
2               Action|Adventure|Thriller  ...                994.0  English   
3                         Action|Thriller  ...               2701.0  English   
4                 Action|Adventure|Sci-Fi  ...                738.0  English   
...                                   ...  ...                  ...      ...   
3848                Drama|Sci-Fi|Thriller  ...                371.0  English   
3849                             Thriller  ...                 35.0  English   
3850  Action|Crime|Drama|Romance|Thriller  ...                130.0  Spanish   
3851                         Comedy|Drama  ...                 14.0  English   
3852                          Documentary  ...                 84.0  English   

          country  content_rating       budget  title_year  \
0             USA           PG-13  237000000.0      2009.0   
1             USA           PG-13  300000000.0      2007.0   
2              UK           PG-13  245000000.0      2015.0   
3             USA           PG-13  250000000.0      2012.0   
4             USA           PG-13  263700000.0      2012.0   
...           ...             ...          ...         ...   
3848          USA           PG-13       7000.0      2004.0   
3849  Philippines       Not Rated       7000.0      2005.0   
3850          USA               R       7000.0      1992.0   
3851          USA       Not Rated       9000.0      2011.0   
3852          USA              PG       1100.0      2004.0   

     actor_2_face

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [180]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3853 entries, 0 to 3852
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      3851 non-null   object 
 1   director_name              3853 non-null   object 
 2   num_critic_for_reviews     3852 non-null   float64
 3   duration                   3852 non-null   float64
 4   director_facebook_likes    3853 non-null   float64
 5   actor_3_facebook_likes     3847 non-null   float64
 6   actor_2_name               3852 non-null   object 
 7   actor_1_facebook_likes     3853 non-null   float64
 8   gross                      3853 non-null   float64
 9   genres                     3853 non-null   object 
 10  actor_1_name               3853 non-null   object 
 11  movie_title                3853 non-null   object 
 12  num_voted_users            3853 non-null   int64  
 13  cast_total_facebook_likes  3853 non-null   int64

In [181]:
movies.describe()

num_critic_for_reviews     duration  director_facebook_likes  \
count             3852.000000  3852.000000              3853.000000   
mean               163.046989   109.921080               784.507397   
std                123.927258    22.741728              3027.395958   
min                  1.000000    34.000000                 0.000000   
25%                 72.000000    95.000000                10.000000   
50%                134.000000   106.000000                58.000000   
75%                221.000000   120.000000               222.000000   
max                813.000000   330.000000             23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count             3847.000000             3853.000000  3.853000e+03   
mean               747.290096             7575.558526  5.096547e+07   
std               1841.623945            15405.383114  6.932033e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                183.000000              721.000000  6.830957e+06   
50%                427.000000             1000.000000  2.790000e+07   
75%                685.000000            12000.000000  6.550000e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count     3.853000e+03                3853.000000           3847.000000   
mean      1.024240e+05               11239.480145              1.378217   
std       1.502942e+05               18922.948757              2.056071   
min       5.000000e+00                   0.000000              0.000000   
25%       1.730900e+04                1817.000000              0.000000   
50%       5.052300e+04                3876.000000              1.000000   
75%       1.241850e+05               15972.000000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews        budget   title_year  \
count           3853.000000  3.853000e+03  3853.000000   
mean             326.720478  4.524817e+07  2003.061511   
std              407.920904  2.234208e+08    10.007168   
min                1.000000  2.180000e+02  1920.000000   
25%              102.000000  1.000000e+07  1999.000000   
50%              203.000000  2.400000e+07  2005.000000   
75%              391.000000  5.000000e+07  2010.000000   
max             5060.000000  1.221550e+10  2016.000000   

       actor_2_facebook_likes   imdb_score  aspect_ratio  movie_facebook_likes  
count             3852.000000  3853.000000   3781.000000           3853.000000  
mean              1959.018432     6.462886      2.109569           9089.546068  
std               4472.171290     1.053843      0.353236          21277.375223  
min                  0.000000     1.600000      1.180000              0.000000  
25%                362.000000     5.900000      1.850000              0.000000  
50%                664.000000     6.600000      2.350000            206.000000  
75%                971.000000     7.200000      2.350000          11000.000000  
max             137000.000000     9.300000     16.000000         349000.000000

In [182]:
movies.shape

(3853, 28)

In [183]:
movies.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

In [184]:
movies.head(10)

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4  Color     Andrew Stanton                   462.0     132.0   
5  Color          Sam Raimi                   392.0     156.0   
6  Color       Nathan Greno                   324.0     100.0   
7  Color        Joss Whedon                   635.0     141.0   
8  Color        David Yates                   375.0     153.0   
9  Color        Zack Snyder                   673.0     183.0   

   director_facebook_likes  actor_3_facebook_likes       actor_2_name  \
0                      0.0                   855.0   Joel David Moore   
1                    563.0                  1000.0      Orlando Bloom   
2                      0.0                   161.0       Rory Kinnear   
3                  22000.0                 23000.0     Christian Bale   
4                    475.0                   530.0    Samantha Morton   
5                      0.0                  4000.0       James Franco   
6                     15.0                   284.0       Donna Murphy   
7                      0.0                 19000.0  Robert Downey Jr.   
8                    282.0                 10000.0   Daniel Radcliffe   
9                      0.0                  2000.0       Lauren Cohan   

   actor_1_facebook_likes        gross  \
0                  1000.0  760505847.0   
1                 40000.0  309404152.0   
2                 11000.0  200074175.0   
3                 27000.0  448130642.0   
4                   640.0   73058679.0   
5                 24000.0  336530303.0   
6                   799.0  200807262.0   
7                 26000.0  458991599.0   
8                 25000.0  301956980.0   
9                 15000.0  330249062.0   

                                              genres  ...  \
0                    Action|Adventure|Fantasy|Sci-Fi  ...   
1                           Action|Adventure|Fantasy  ...   
2                          Action|Adventure|Thriller  ...   
3                                    Action|Thriller  ...   
4                            Action|Adventure|Sci-Fi  ...   
5                           Action|Adventure|Romance  ...   
6  Adventure|Animation|Comedy|Family|Fantasy|Musi...  ...   
7                            Action|Adventure|Sci-Fi  ...   
8                   Adventure|Family|Fantasy|Mystery  ...   
9                            Action|Adventure|Sci-Fi  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                738.0  English      USA           PG-13  263700000.0   
5               1902.0  English      USA           PG-13  258000000.0   
6                387.0  English      USA              PG  260000000.0   
7               1117.0  English      USA           PG-13  250000000.0   
8                973.0  English       UK              PG  250000000.0   
9               3018.0  English      USA           PG-13  250000000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4      2012.0                  632.0        6.6          2.35   
5      2007.0                11000.0        6.2          2.35   
6      2010.0                  553.0        7.8          1.85   
7      2015.0           

In [185]:
movies.tail(10)

color     director_name  num_critic_for_reviews  duration  \
3843  Color        Daryl Wein                    22.0      88.0   
3844  Color       John Waters                    73.0     108.0   
3845  Color   Olivier Assayas                    81.0     110.0   
3846  Color      Jafar Panahi                    64.0      90.0   
3847  Color  Kiyoshi Kurosawa                    78.0     111.0   
3848  Color     Shane Carruth                   143.0      77.0   
3849  Color  Neill Dela Llana                    35.0      80.0   
3850  Color  Robert Rodriguez                    56.0      81.0   
3851  Color      Edward Burns                    14.0      95.0   
3852  Color          Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
3843                     38.0                   211.0       Heather Burns   
3844                      0.0                   105.0          Mink Stole   
3845                    107.0                    45.0      Béatrice Dalle   
3846                    397.0                     0.0   Nargess Mamizadeh   
3847                     62.0                     6.0       Anna Nakagawa   
3848                    291.0                     8.0      David Sullivan   
3849                      0.0                     0.0     Edgar Tancangco   
3850                      0.0                     6.0     Peter Marquardt   
3851                      0.0                   133.0  Caitlin FitzGerald   
3852                     16.0                    16.0    Brian Herzlinger   

      actor_1_facebook_likes      gross                               genres  \
3843                   331.0    76382.0                              Romance   
3844                   462.0   180483.0                  Comedy|Crime|Horror   
3845                   576.0   136007.0                  Drama|Music|Romance   
3846                     5.0   673780.0                                Drama   
3847                    89.0    94596.0        Crime|Horror|Mystery|Thriller   
3848                   291.0   424760.0                Drama|Sci-Fi|Thriller   
3849                     0.0    70071.0                             Thriller   
3850                   121.0  2040920.0  Action|Crime|Drama|Romance|Thriller   
3851                   296.0     4584.0                         Comedy|Drama   
3852                    86.0    85222.0                          Documentary   

      ... num_user_for_reviews  language      country  content_rating  \
3843  ...                  8.0   English          USA             NaN   
3844  ...                183.0   English          USA           NC-17   
3845  ...                 39.0    French       France               R   
3846  ...                 26.0   Persian         Iran       Not Rated   
3847  ...                 50.0  Japanese        Japan             NaN   
3848  ...                371.0   English          USA           PG-13   
3849  ...                 35.0   English  Philippines       Not Rated   
3850  ...                130.0   Spanish          USA               R   
3851  ...                 14.0   English          USA       Not Rated   
3852  ...                 84.0   English          USA              PG   

         budget  title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
3843    15000.0      2009.0                  212.0        6.2          2.35   
3844    10000.0      1972.0                  143.0        6.1          1.37   
3845     4500.0      2004.0                  133.0        6.9          2.35   
3846    10000.0      2000.0                    0.0        7.5          1.85   
3847  1000000.0      1997.0                   13.0        7.4          1.85   
3848     7000.0      2004.0                   45.0        7.0          1.85   
3849     7000.0      2005.0                    0.0        6.3           NaN   
3850     7000.0      1992.0                   20.0        6.9          1.37   
3851     9000.0      201

How many rows and columns are present in the dataframe?

In [186]:
movies.shape

(3853, 28)

3853 rows, 28 columns

How many columns have null values present in them?

In [187]:
movies.isnull().any()

color                         True
director_name                False
num_critic_for_reviews        True
duration                      True
director_facebook_likes      False
actor_3_facebook_likes        True
actor_2_name                  True
actor_1_facebook_likes       False
gross                        False
genres                       False
actor_1_name                 False
movie_title                  False
num_voted_users              False
cast_total_facebook_likes    False
actor_3_name                  True
facenumber_in_poster          True
plot_keywords                 True
movie_imdb_link              False
num_user_for_reviews         False
language                      True
country                      False
content_rating                True
budget                       False
title_year                   False
actor_2_facebook_likes        True
imdb_score                   False
aspect_ratio                  True
movie_facebook_likes         False
dtype: bool

In [188]:
columns_with_null = movies.columns[movies.isnull().any()]
print(columns_with_null)

Index(['color', 'num_critic_for_reviews', 'duration', 'actor_3_facebook_likes',
       'actor_2_name', 'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'language', 'content_rating', 'actor_2_facebook_likes', 'aspect_ratio'],
      dtype='object')


In [189]:
len(columns_with_null)

12

There are 12 columns with null values.

### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this project, I will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. Some of the columns in this dataframe are not required. So I will be dropping the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [190]:
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4  Color     Andrew Stanton                   462.0     132.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    475.0                   530.0   Samantha Morton   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   640.0   73058679.0          Action|Adventure|Sci-Fi  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                738.0  English      USA           PG-13  263700000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4      2012.0                  632.0        6.6          2.35   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                24000  

[5 rows x 28 columns]

In [191]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3853 entries, 0 to 3852
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      3851 non-null   object 
 1   director_name              3853 non-null   object 
 2   num_critic_for_reviews     3852 non-null   float64
 3   duration                   3852 non-null   float64
 4   director_facebook_likes    3853 non-null   float64
 5   actor_3_facebook_likes     3847 non-null   float64
 6   actor_2_name               3852 non-null   object 
 7   actor_1_facebook_likes     3853 non-null   float64
 8   gross                      3853 non-null   float64
 9   genres                     3853 non-null   object 
 10  actor_1_name               3853 non-null   object 
 11  movie_title                3853 non-null   object 
 12  num_voted_users            3853 non-null   int64  
 13  cast_total_facebook_likes  3853 non-null   int64

In [192]:
# Drop the columns that are not necessary
drop_columns = ['color', 'director_facebook_likes','actor_1_facebook_likes', 'actor_2_facebook_likes', 
                'actor_3_facebook_likes', 'actor_2_name', 'cast_total_facebook_likes', 'actor_3_name',
                'duration', 'facenumber_in_poster', 'content_rating', 'country', 'movie_imdb_link',
                'aspect_ratio', 'plot_keywords']
new_movies = movies.drop(columns = drop_columns)
new_movies.head()

director_name  num_critic_for_reviews        gross  \
0      James Cameron                   723.0  760505847.0   
1     Gore Verbinski                   302.0  309404152.0   
2         Sam Mendes                   602.0  200074175.0   
3  Christopher Nolan                   813.0  448130642.0   
4     Andrew Stanton                   462.0   73058679.0   

                            genres     actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1         Action|Adventure|Fantasy      Johnny Depp   
2        Action|Adventure|Thriller  Christoph Waltz   
3                  Action|Thriller        Tom Hardy   
4          Action|Adventure|Sci-Fi     Daryl Sabara   

                                 movie_title  num_voted_users  \
0                                    Avatar            886204   
1  Pirates of the Caribbean: At World's End            471220   
2                                   Spectre            275868   
3                     The Dark Knight Rises           1144337   
4                               John Carter            212204   

   num_user_for_reviews language       budget  title_year  imdb_score  \
0                3054.0  English  237000000.0      2009.0         7.9   
1                1238.0  English  300000000.0      2007.0         7.1   
2                 994.0  English  245000000.0      2015.0         6.8   
3                2701.0  English  250000000.0      2012.0         8.5   
4                 738.0  English  263700000.0      2012.0         6.6   

   movie_facebook_likes  
0                 33000  
1                     0  
2                 85000  
3                164000  
4                 24000

In [193]:
new_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3853 entries, 0 to 3852
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   director_name           3853 non-null   object 
 1   num_critic_for_reviews  3852 non-null   float64
 2   gross                   3853 non-null   float64
 3   genres                  3853 non-null   object 
 4   actor_1_name            3853 non-null   object 
 5   movie_title             3853 non-null   object 
 6   num_voted_users         3853 non-null   int64  
 7   num_user_for_reviews    3853 non-null   float64
 8   language                3849 non-null   object 
 9   budget                  3853 non-null   float64
 10  title_year              3853 non-null   float64
 11  imdb_score              3853 non-null   float64
 12  movie_facebook_likes    3853 non-null   int64  
dtypes: float64(6), int64(2), object(5)
memory usage: 391.4+ KB


What is the count of columns in the new dataframe?

In [194]:
new_movies.shape

(3853, 13)

New dataframe has 13 columns.

**Subtask 2.2: Inspect Null values**

Since, there are null values in multiple columns of the dataframe. Find out the percentage of null values in each column of the dataframe 'movies'. 

In [195]:
movies_null = new_movies[new_movies.columns[new_movies.isnull().any()]]
movies_null

num_critic_for_reviews language
0                      723.0  English
1                      302.0  English
2                      602.0  English
3                      813.0  English
4                      462.0  English
...                      ...      ...
3848                   143.0  English
3849                    35.0  English
3850                    56.0  Spanish
3851                    14.0  English
3852                    43.0  English

[3853 rows x 2 columns]

In [196]:
movies_null.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3853 entries, 0 to 3852
Data columns (total 2 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   num_critic_for_reviews  3852 non-null   float64
 1   language                3849 non-null   object 
dtypes: float64(1), object(1)
memory usage: 60.3+ KB


In [197]:
# count percentage of null values
null_percentage = round((movies_null.isnull().sum() / 3853)*100,2) 
null_percentage

num_critic_for_reviews    0.03
language                  0.10
dtype: float64

In [198]:
percentage_null_df = pd.DataFrame(null_percentage, columns=["Percentage of Null Values (%)"])
percentage_null_df.sort_values("Percentage of Null Values (%)")

Percentage of Null Values (%)
num_critic_for_reviews                           0.03
language                                         0.10

Which column has the highest percentage of null values?

In [199]:
percentage_null_df['Percentage of Null Values (%)'].max()

0.1

Language has the highest percentage of null values.

In [200]:
new_movies['language'].groupby(new_movies['language']).count().sort_values(ascending=False)

language
English       3671
French          37
Spanish         26
Mandarin        14
German          13
Japanese        12
Hindi           10
Cantonese        8
Italian          7
Portuguese       5
Korean           5
Norwegian        4
Dutch            3
Persian          3
Danish           3
Thai             3
Aboriginal       2
Indonesian       2
Dari             2
Hebrew           2
Czech            1
Vietnamese       1
Aramaic          1
Telugu           1
Swedish          1
Bosnian          1
Russian          1
Romanian         1
Arabic           1
Icelandic        1
Dzongkha         1
Mongolian        1
Maya             1
Filipino         1
Kazakh           1
Hungarian        1
Zulu             1
Name: language, dtype: int64

**Subtask 2.3: Fill NaN values**

Since the `language` column has some NaN values. Since, English the most commonly used in the film industry, especially in this dataset, I will replace the null values with `'English'`.

In [201]:
# replace the NaN values in the 'language' column with English

new_movies['language'] = new_movies['language'].fillna("English")
new_movies.isnull().any()

director_name             False
num_critic_for_reviews     True
gross                     False
genres                    False
actor_1_name              False
movie_title               False
num_voted_users           False
num_user_for_reviews      False
language                  False
budget                    False
title_year                False
imdb_score                False
movie_facebook_likes      False
dtype: bool

What is the count of movies made in English language after replacing the NaN values with English?

In [202]:
new_movies.shape

(3853, 13)

In [203]:
print("Count of movies in English: ",new_movies[new_movies['language'] == 'English'].shape[0])

Count of movies in English:  3675


### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [204]:
new_movies.head()

director_name  num_critic_for_reviews        gross  \
0      James Cameron                   723.0  760505847.0   
1     Gore Verbinski                   302.0  309404152.0   
2         Sam Mendes                   602.0  200074175.0   
3  Christopher Nolan                   813.0  448130642.0   
4     Andrew Stanton                   462.0   73058679.0   

                            genres     actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1         Action|Adventure|Fantasy      Johnny Depp   
2        Action|Adventure|Thriller  Christoph Waltz   
3                  Action|Thriller        Tom Hardy   
4          Action|Adventure|Sci-Fi     Daryl Sabara   

                                 movie_title  num_voted_users  \
0                                    Avatar            886204   
1  Pirates of the Caribbean: At World's End            471220   
2                                   Spectre            275868   
3                     The Dark Knight Rises           1144337   
4                               John Carter            212204   

   num_user_for_reviews language       budget  title_year  imdb_score  \
0                3054.0  English  237000000.0      2009.0         7.9   
1                1238.0  English  300000000.0      2007.0         7.1   
2                 994.0  English  245000000.0      2015.0         6.8   
3                2701.0  English  250000000.0      2012.0         8.5   
4                 738.0  English  263700000.0      2012.0         6.6   

   movie_facebook_likes  
0                 33000  
1                     0  
2                 85000  
3                164000  
4                 24000

In [205]:
# divide budget and gross by 1 million
new_movies['budget'] = new_movies['budget'] / 1000000  
new_movies['gross'] = new_movies['gross'] / 1000000
new_movies.head()

director_name  num_critic_for_reviews       gross  \
0      James Cameron                   723.0  760.505847   
1     Gore Verbinski                   302.0  309.404152   
2         Sam Mendes                   602.0  200.074175   
3  Christopher Nolan                   813.0  448.130642   
4     Andrew Stanton                   462.0   73.058679   

                            genres     actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1         Action|Adventure|Fantasy      Johnny Depp   
2        Action|Adventure|Thriller  Christoph Waltz   
3                  Action|Thriller        Tom Hardy   
4          Action|Adventure|Sci-Fi     Daryl Sabara   

                                 movie_title  num_voted_users  \
0                                    Avatar            886204   
1  Pirates of the Caribbean: At World's End            471220   
2                                   Spectre            275868   
3                     The Dark Knight Rises           1144337   
4                               John Carter            212204   

   num_user_for_reviews language  budget  title_year  imdb_score  \
0                3054.0  English   237.0      2009.0         7.9   
1                1238.0  English   300.0      2007.0         7.1   
2                 994.0  English   245.0      2015.0         6.8   
3                2701.0  English   250.0      2012.0         8.5   
4                 738.0  English   263.7      2012.0         6.6   

   movie_facebook_likes  
0                 33000  
1                     0  
2                 85000  
3                164000  
4                 24000

**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. 
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [206]:
# assign new column profit
new_movies['profit'] = new_movies['gross'] - new_movies['budget']
new_movies.head()

director_name  num_critic_for_reviews       gross  \
0      James Cameron                   723.0  760.505847   
1     Gore Verbinski                   302.0  309.404152   
2         Sam Mendes                   602.0  200.074175   
3  Christopher Nolan                   813.0  448.130642   
4     Andrew Stanton                   462.0   73.058679   

                            genres     actor_1_name  \
0  Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1         Action|Adventure|Fantasy      Johnny Depp   
2        Action|Adventure|Thriller  Christoph Waltz   
3                  Action|Thriller        Tom Hardy   
4          Action|Adventure|Sci-Fi     Daryl Sabara   

                                 movie_title  num_voted_users  \
0                                    Avatar            886204   
1  Pirates of the Caribbean: At World's End            471220   
2                                   Spectre            275868   
3                     The Dark Knight Rises           1144337   
4                               John Carter            212204   

   num_user_for_reviews language  budget  title_year  imdb_score  \
0                3054.0  English   237.0      2009.0         7.9   
1                1238.0  English   300.0      2007.0         7.1   
2                 994.0  English   245.0      2015.0         6.8   
3                2701.0  English   250.0      2012.0         8.5   
4                 738.0  English   263.7      2012.0         6.6   

   movie_facebook_likes      profit  
0                 33000  523.505847  
1                     0    9.404152  
2                 85000  -44.925825  
3                164000  198.130642  
4                 24000 -190.641321

In [207]:
# sort
new_movies.sort_values('profit', ascending=False).head(10)

director_name  num_critic_for_reviews       gross  \
0         James Cameron                   723.0  760.505847   
28      Colin Trevorrow                   644.0  652.177271   
25        James Cameron                   315.0  658.672302   
2704       George Lucas                   282.0  460.935665   
2748   Steven Spielberg                   215.0  434.949459   
16          Joss Whedon                   703.0  623.279547   
482        Roger Allers                   186.0  422.783777   
230        George Lucas                   320.0  474.544677   
64    Christopher Nolan                   645.0  533.316061   
419           Gary Ross                   673.0  407.999255   

                                        genres         actor_1_name  \
0              Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
28            Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
25                               Drama|Romance    Leonardo DiCaprio   
2704           Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
2748                             Family|Sci-Fi         Henry Thomas   
16                     Action|Adventure|Sci-Fi      Chris Hemsworth   
482   Adventure|Animation|Drama|Family|Musical    Matthew Broderick   
230            Action|Adventure|Fantasy|Sci-Fi      Natalie Portman   
64                 Action|Crime|Drama|Thriller       Christian Bale   
419            Adventure|Drama|Sci-Fi|Thriller    Jennifer Lawrence   

                                     movie_title  num_voted_users  \
0                                        Avatar            886204   
28                               Jurassic World            418214   
25                                      Titanic            793059   
2704         Star Wars: Episode IV - A New Hope            911097   
2748                 E.T. the Extra-Terrestrial            281842   
16                                 The Avengers            995415   
482                               The Lion King            644348   
230   Star Wars: Episode I - The Phantom Menace            534658   
64                              The Dark Knight           1676169   
419                            The Hunger Games            701607   

      num_user_for_reviews language  budget  title_year  imdb_score  \
0                   3054.0  English   237.0      2009.0         7.9   
28                  1290.0  English   150.0      2015.0         7.0   
25                  2528.0  English   200.0      1997.0         7.7   
2704                1470.0  English    11.0      1977.0         8.7   
2748                 515.0  English    10.5      1982.0         7.9   
16                  1722.0  English   220.0      2012.0         8.1   
482                  656.0  English    45.0      1994.0         8.5   
230                 3597.0  English   115.0      1999.0         6.5   
64                  4667.0  English   185.0      2008.0         9.0   
419                 1959.0  English    78.0      2012.0         7.3   

      movie_facebook_likes      profit  
0                    33000  523.505847  
28                  150000  502.177271  
25                   26000  458.672302  
2704                 33000  449.935665  
2748                 34000  424.449459  
16                  123000  403.279547  
482                  17000  377.783777  
230                  13000  359.544677  
64                   37000  348.316061  
419                 140000  329.999255

In [208]:
# get the top 10 profiting movies

top10_movies = new_movies.sort_values('profit', ascending=False).head(10)
top10_movies

director_name  num_critic_for_reviews       gross  \
0         James Cameron                   723.0  760.505847   
28      Colin Trevorrow                   644.0  652.177271   
25        James Cameron                   315.0  658.672302   
2704       George Lucas                   282.0  460.935665   
2748   Steven Spielberg                   215.0  434.949459   
16          Joss Whedon                   703.0  623.279547   
482        Roger Allers                   186.0  422.783777   
230        George Lucas                   320.0  474.544677   
64    Christopher Nolan                   645.0  533.316061   
419           Gary Ross                   673.0  407.999255   

                                        genres         actor_1_name  \
0              Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
28            Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
25                               Drama|Romance    Leonardo DiCaprio   
2704           Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
2748                             Family|Sci-Fi         Henry Thomas   
16                     Action|Adventure|Sci-Fi      Chris Hemsworth   
482   Adventure|Animation|Drama|Family|Musical    Matthew Broderick   
230            Action|Adventure|Fantasy|Sci-Fi      Natalie Portman   
64                 Action|Crime|Drama|Thriller       Christian Bale   
419            Adventure|Drama|Sci-Fi|Thriller    Jennifer Lawrence   

                                     movie_title  num_voted_users  \
0                                        Avatar            886204   
28                               Jurassic World            418214   
25                                      Titanic            793059   
2704         Star Wars: Episode IV - A New Hope            911097   
2748                 E.T. the Extra-Terrestrial            281842   
16                                 The Avengers            995415   
482                               The Lion King            644348   
230   Star Wars: Episode I - The Phantom Menace            534658   
64                              The Dark Knight           1676169   
419                            The Hunger Games            701607   

      num_user_for_reviews language  budget  title_year  imdb_score  \
0                   3054.0  English   237.0      2009.0         7.9   
28                  1290.0  English   150.0      2015.0         7.0   
25                  2528.0  English   200.0      1997.0         7.7   
2704                1470.0  English    11.0      1977.0         8.7   
2748                 515.0  English    10.5      1982.0         7.9   
16                  1722.0  English   220.0      2012.0         8.1   
482                  656.0  English    45.0      1994.0         8.5   
230                 3597.0  English   115.0      1999.0         6.5   
64                  4667.0  English   185.0      2008.0         9.0   
419                 1959.0  English    78.0      2012.0         7.3   

      movie_facebook_likes      profit  
0                    33000  523.505847  
28                  150000  502.177271  
25                   26000  458.672302  
2704                 33000  449.935665  
2748                 34000  424.449459  
16                  123000  403.279547  
482                  17000  377.783777  
230                  13000  359.544677  
64                   37000  348.316061  
419                 140000  329.999255

- Six out of the top 10 films with highest profits, belong to the action genre.
- Movie with the highest profit is Avatar by James Cameron.
- There are two movies directed by James Cameron in the top 10 most profitable movies.

Which movie is ranked 5th from the top in the list obtained?

In [209]:
top10_movies[4:5]

director_name  num_critic_for_reviews       gross         genres  \
2748  Steven Spielberg                   215.0  434.949459  Family|Sci-Fi   

      actor_1_name                  movie_title  num_voted_users  \
2748  Henry Thomas  E.T. the Extra-Terrestrial            281842   

      num_user_for_reviews language  budget  title_year  imdb_score  \
2748                 515.0  English    10.5      1982.0         7.9   

      movie_facebook_likes      profit  
2748                 34000  424.449459

Movie ranked 5th from the top is E.T. the Extra-Terrestrial.

**Subtask 3.3: Find IMDb Top 200**

Create a new dataframe `IMDb_Top_200` and store the top 200 movies with the highest IMDb Rating (column: `imdb_score`). And only get movies with the `num_voted_users` is greater than 25,000. 

Add `Rank` column containing the values 1 to 200 indicating the ranks of the corresponding films.

In [210]:
# New df 'IMDb_Top_200', sort values in descending order

IMDb_Top_200 = new_movies.sort_values('imdb_score', ascending=False)
IMDb_Top_200.head()

director_name  num_critic_for_reviews       gross  \
1795        Frank Darabont                   199.0   28.341469   
3016  Francis Ford Coppola                   208.0  134.821952   
2543  Francis Ford Coppola                   149.0   57.300000   
64       Christopher Nolan                   645.0  533.316061   
325          Peter Jackson                   328.0  377.019252   

                              genres    actor_1_name  \
1795                     Crime|Drama  Morgan Freeman   
3016                     Crime|Drama       Al Pacino   
2543                     Crime|Drama  Robert De Niro   
64       Action|Crime|Drama|Thriller  Christian Bale   
325   Action|Adventure|Drama|Fantasy   Orlando Bloom   

                                         movie_title  num_voted_users  \
1795                       The Shawshank Redemption           1689764   
3016                                  The Godfather           1155770   
2543                         The Godfather: Part II            790926   
64                                  The Dark Knight           1676169   
325   The Lord of the Rings: The Return of the King           1215718   

      num_user_for_reviews language  budget  title_year  imdb_score  \
1795                4144.0  English    25.0      1994.0         9.3   
3016                2238.0  English     6.0      1972.0         9.2   
2543                 650.0  English    13.0      1974.0         9.0   
64                  4667.0  English   185.0      2008.0         9.0   
325                 3189.0  English    94.0      2003.0         8.9   

      movie_facebook_likes      profit  
1795                108000    3.341469  
3016                 43000  128.821952  
2543                 14000   44.300000  
64                   37000  348.316061  
325                  16000  283.019252

In [211]:
# only get rows with num_voted_users > 25000
IMDb_Top_200 = IMDb_Top_200[IMDb_Top_200.num_voted_users > 25000]
IMDb_Top_200

director_name  num_critic_for_reviews       gross  \
1795        Frank Darabont                   199.0   28.341469   
3016  Francis Ford Coppola                   208.0  134.821952   
2543  Francis Ford Coppola                   149.0   57.300000   
64       Christopher Nolan                   645.0  533.316061   
325          Peter Jackson                   328.0  377.019252   
...                    ...                     ...         ...   
2019       Jason Friedberg                   112.0   39.737645   
305      Lawrence Guterman                    78.0   17.010646   
2085       Jason Friedberg                   111.0   14.174654   
2111             Bob Clark                    32.0    9.109322   
2540            Jon M. Chu                    84.0   73.000942   

                              genres     actor_1_name  \
1795                     Crime|Drama   Morgan Freeman   
3016                     Crime|Drama        Al Pacino   
2543                     Crime|Drama   Robert De Niro   
64       Action|Crime|Drama|Thriller   Christian Bale   
325   Action|Adventure|Drama|Fantasy    Orlando Bloom   
...                              ...              ...   
2019                Adventure|Comedy  David Carradine   
305            Comedy|Family|Fantasy    Jamie Kennedy   
2085                          Comedy   Carmen Electra   
2111            Comedy|Family|Sci-Fi       Scott Baio   
2540               Documentary|Music    Usher Raymond   

                                         movie_title  num_voted_users  \
1795                       The Shawshank Redemption           1689764   
3016                                  The Godfather           1155770   
2543                         The Godfather: Part II            790926   
64                                  The Dark Knight           1676169   
325   The Lord of the Rings: The Return of the King           1215718   
...                                              ...              ...   
2019                                     Epic Movie             89687   
305                                 Son of the Mask             40751   
2085                                 Disaster Movie             74945   
2111                   Superbabies: Baby Geniuses 2             25371   
2540                 Justin Bieber: Never Say Never             74351   

      num_user_for_reviews language  budget  title_year  imdb_score  \
1795                4144.0  English    25.0      1994.0         9.3   
3016                2238.0  English     6.0      1972.0         9.2   
2543                 650.0  English    13.0      1974.0         9.0   
64                  4667.0  English   185.0      2008.0         9.0   
325                 3189.0  English    94.0      2003.0         8.9   
...                    ...      ...     ...         ...         ...   
2019                 666.0  English    20.0      2007.0         2.3   
305                  239.0  English    84.0      2005.0         2.2   
2085                 359.0  English    25.0      2008.0         1.9   
2111                 129.0  English    20.0      2004.0         1.9   
2540                 233.0  English    13.0      2011.0         1.6   

      movie_facebook_likes      profit  
1795                108000    3.341469  
3016                 43000  128.821952  
2543                 14000   44.300000  
64                   37000  348.316061  
325                  16000  283.019252  
...                    ...         ...  
2019                     0   19.737645  
305                    881  -66.989354  
2085                     0  -10.825346  
2111                     0  -10.890678  
2540                 62000   60.000942  

[2610 rows x 14 columns]

In [212]:
# return only the first 200 rows
IMDb_Top_200 = IMDb_Top_200[0:200]
IMDb_Top_200

director_name  num_critic_for_reviews       gross  \
1795        Frank Darabont                   199.0   28.341469   
3016  Francis Ford Coppola                   208.0  134.821952   
2543  Francis Ford Coppola                   149.0   57.300000   
64       Christopher Nolan                   645.0  533.316061   
325          Peter Jackson                   328.0  377.019252   
...                    ...                     ...         ...   
1622          Ron Clements                   124.0  217.350219   
293           Edward Zwick                   166.0   57.366262   
2304            Peter Weir                    96.0   95.860116   
2360            Tom Hooper                   479.0  138.795342   
155            J.J. Abrams                   518.0  257.704099   

                                                 genres       actor_1_name  \
1795                                        Crime|Drama     Morgan Freeman   
3016                                        Crime|Drama          Al Pacino   
2543                                        Crime|Drama     Robert De Niro   
64                          Action|Crime|Drama|Thriller     Christian Bale   
325                      Action|Adventure|Drama|Fantasy      Orlando Bloom   
...                                                 ...                ...   
1622  Adventure|Animation|Comedy|Family|Fantasy|Musi...     Robin Williams   
293                            Adventure|Drama|Thriller  Leonardo DiCaprio   
2304                                       Comedy|Drama     Robin Williams   
2360                    Biography|Drama|History|Romance        Colin Firth   
155                             Action|Adventure|Sci-Fi    Chris Hemsworth   

                                         movie_title  num_voted_users  \
1795                       The Shawshank Redemption           1689764   
3016                                  The Godfather           1155770   
2543                         The Godfather: Part II            790926   
64                                  The Dark Knight           1676169   
325   The Lord of the Rings: The Return of the King           1215718   
...                                              ...              ...   
1622                                        Aladdin            260939   
293                                   Blood Diamond            400292   
2304                             Dead Poets Society            277451   
2360                              The King's Speech            503631   
155                                       Star Trek            504419   

      num_user_for_reviews language  budget  title_year  imdb_score  \
1795                4144.0  English    25.0      1994.0         9.3   
3016                2238.0  English     6.0      1972.0         9.2   
2543                 650.0  English    13.0      1974.0         9.0   
64                  4667.0  English   185.0      2008.0         9.0   
325                 3189.0  English    94.0      2003.0         8.9   
...                    ...      ...     ...         ...         ...   
1622                 244.0  English    28.0      1992.0         8.0   
293                  657.0  English   100.0      2006.0         8.0   
2304                 491.0  English    16.4      1989.0         8.0   
2360                 636.0  English    15.0      2010.0         8.0   
155                 1559.0  English   150.0      2009.0         8.0   

      movie_facebook_likes      profit  
1795                108000    3.341469  
3016                 43000  128.821952  
2543                 14000   44.300000  
64                   37000  348.316061  
325                  16000  283.019252  
...                    ...         ...  
1622                     0  189.350219  
293                  14000  -42.633738  
2304                 23000   79.460116  
2360                 64000  123.795342  
155                  19000  107.704099  

[200 rows x 14 columns]

In [213]:
# create a new column called rank and move it to the first column

IMDb_Top_200 = IMDb_Top_200.copy() # make copy of the dataframe to prevent setting warning
IMDb_Top_200['rank'] = np.arange(1,201)
IMDb_Top_200.insert(0, 'Rank', IMDb_Top_200['rank'])
IMDb_Top_200.drop(columns=['rank'], inplace=True)
IMDb_Top_200.head(10)

Rank         director_name  num_critic_for_reviews       gross  \
1795     1        Frank Darabont                   199.0   28.341469   
3016     2  Francis Ford Coppola                   208.0  134.821952   
2543     3  Francis Ford Coppola                   149.0   57.300000   
64       4     Christopher Nolan                   645.0  533.316061   
325      5         Peter Jackson                   328.0  377.019252   
3607     6          Sergio Leone                   181.0    6.100000   
2938     7     Quentin Tarantino                   215.0  107.930000   
1737     8      Steven Spielberg                   174.0   96.067179   
94       9     Christopher Nolan                   642.0  292.568851   
646     10         David Fincher                   315.0   37.023395   

                                genres       actor_1_name  \
1795                       Crime|Drama     Morgan Freeman   
3016                       Crime|Drama          Al Pacino   
2543                       Crime|Drama     Robert De Niro   
64         Action|Crime|Drama|Thriller     Christian Bale   
325     Action|Adventure|Drama|Fantasy      Orlando Bloom   
3607                           Western     Clint Eastwood   
2938                       Crime|Drama       Bruce Willis   
1737           Biography|Drama|History        Liam Neeson   
94    Action|Adventure|Sci-Fi|Thriller  Leonardo DiCaprio   
646                              Drama          Brad Pitt   

                                         movie_title  num_voted_users  \
1795                       The Shawshank Redemption           1689764   
3016                                  The Godfather           1155770   
2543                         The Godfather: Part II            790926   
64                                  The Dark Knight           1676169   
325   The Lord of the Rings: The Return of the King           1215718   
3607                 The Good, the Bad and the Ugly            503509   
2938                                   Pulp Fiction           1324680   
1737                               Schindler's List            865020   
94                                        Inception           1468200   
646                                      Fight Club           1347461   

      num_user_for_reviews language  budget  title_year  imdb_score  \
1795                4144.0  English    25.0      1994.0         9.3   
3016                2238.0  English     6.0      1972.0         9.2   
2543                 650.0  English    13.0      1974.0         9.0   
64                  4667.0  English   185.0      2008.0         9.0   
325                 3189.0  English    94.0      2003.0         8.9   
3607                 780.0  Italian     1.2      1966.0         8.9   
2938                2195.0  English     8.0      1994.0         8.9   
1737                1273.0  English    22.0      1993.0         8.9   
94                  2803.0  English   160.0      2010.0         8.8   
646                 2968.0  English    63.0      1999.0         8.8   

      movie_facebook_likes      profit  
1795                108000    3.341469  
3016                 43000  128.821952  
2543                 14000   44.300000  
64                   37000  348.316061  
325                  16000  283.019252  
3607                 20000    4.900000  
2938                 45000   99.930000  
1737                 41000   74.067179  
94                  175000  132.568851  
646                  48000  -25.976605

In [214]:
# return top 10 profits from the imdb top 200
IMDb_Top_200.sort_values('profit', ascending=False).head(10)

Rank      director_name  num_critic_for_reviews       gross  \
2704    18       George Lucas                   282.0  460.935665   
16     119        Joss Whedon                   703.0  623.279547   
482     46       Roger Allers                   186.0  422.783777   
64       4  Christopher Nolan                   645.0  533.316061   
767    108         Tim Miller                   579.0  363.024263   
659    109   Steven Spielberg                   308.0  356.784000   
324    102     Andrew Stanton                   301.0  380.838870   
325      5      Peter Jackson                   328.0  377.019252   
1435    49   Richard Marquand                   197.0  309.125409   
787     12    Robert Zemeckis                   149.0  329.691196   

                                        genres       actor_1_name  \
2704           Action|Adventure|Fantasy|Sci-Fi      Harrison Ford   
16                     Action|Adventure|Sci-Fi    Chris Hemsworth   
482   Adventure|Animation|Drama|Family|Musical  Matthew Broderick   
64                 Action|Crime|Drama|Thriller     Christian Bale   
767     Action|Adventure|Comedy|Romance|Sci-Fi      Ryan Reynolds   
659                  Adventure|Sci-Fi|Thriller       Wayne Knight   
324          Adventure|Animation|Comedy|Family    Alexander Gould   
325             Action|Adventure|Drama|Fantasy      Orlando Bloom   
1435           Action|Adventure|Fantasy|Sci-Fi      Harrison Ford   
787                               Comedy|Drama          Tom Hanks   

                                         movie_title  num_voted_users  \
2704             Star Wars: Episode IV - A New Hope            911097   
16                                     The Avengers            995415   
482                                   The Lion King            644348   
64                                  The Dark Knight           1676169   
767                                        Deadpool            479047   
659                                   Jurassic Park            613473   
324                                    Finding Nemo            692482   
325   The Lord of the Rings: The Return of the King           1215718   
1435     Star Wars: Episode VI - Return of the Jedi            681857   
787                                    Forrest Gump           1251222   

      num_user_for_reviews language  budget  title_year  imdb_score  \
2704                1470.0  English    11.0      1977.0         8.7   
16                  1722.0  English   220.0      2012.0         8.1   
482                  656.0  English    45.0      1994.0         8.5   
64                  4667.0  English   185.0      2008.0         9.0   
767                 1058.0  English    58.0      2016.0         8.1   
659                  895.0  English    63.0      1993.0         8.1   
324                  866.0  English    94.0      2003.0         8.2   
325                 3189.0  English    94.0      2003.0         8.9   
1435                 647.0  English    32.5      1983.0         8.4   
787                 1398.0  English    55.0      1994.0         8.8   

      movie_facebook_likes      profit  
2704                 33000  449.935665  
16                  123000  403.279547  
482                  17000  377.783777  
64                   37000  348.316061  
767                 117000  305.024263  
659                  19000  293.784000  
324                  11000  286.838870  
325                  16000  283.019252  
1435                 14000  276.625409  
787                  59000  274.691196

- The top 5 movies with the highest IMDB score belongs to crime and action genre.
- Although at rank 18th, Star Wars: Episode IV - A New Hope by George Lucas, secures first place in terms of profit. Likely due to the big fanbase of the Star Wars series.
- The top 10 highest profits belongs to mostly action and adventure genre.
- The success and profitability of a movie are likely influenced by factors such as belonging to the action genre and having a devoted fanbase like Marvel and Star Wars.

Suppose movies are divided into 5 buckets based on the IMDb ratings: 
-  7.5 to 8
-  8 to 8.5
-  8.5 to 9
-  9 to 9.5
-  9.5 to 10

Which bucket holds the maximum number of movies from 'IMDb_Top_200'?

In [215]:
# filter values by IMDB rating buckets
print("Number of movies by IMDb rating")
print("7.5 to 8 = " , IMDb_Top_200[(IMDb_Top_200['imdb_score'] >= 7.5) & (IMDb_Top_200['imdb_score'] <= 8)].shape[0])
print("8 to 8.5 = " , IMDb_Top_200[(IMDb_Top_200['imdb_score'] >= 8) & (IMDb_Top_200['imdb_score'] <= 8.5)].shape[0])
print("8.5 to 9 = " , IMDb_Top_200[(IMDb_Top_200['imdb_score'] >= 8.5) & (IMDb_Top_200['imdb_score'] <= 9)].shape[0])
print("9 to 9.5 = " , IMDb_Top_200[(IMDb_Top_200['imdb_score'] >= 9) & (IMDb_Top_200['imdb_score'] <= 9.5)].shape[0])
print("9.5 to 10 = " , IMDb_Top_200[(IMDb_Top_200['imdb_score'] >= 9.5) & (IMDb_Top_200['imdb_score'] <= 10.5)].shape[0])

print("Bucket with Max number of movies is 8 to 8.5.")

Number of movies by IMDb rating
7.5 to 8 =  48
8 to 8.5 =  172
8.5 to 9 =  44
9 to 9.5 =  4
9.5 to 10 =  0
Bucket with Max number of movies is 8 to 8.5.


**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. From the `new_movies` dataframe, create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [216]:
# Meryl Streep
Meryl_Streep = new_movies[new_movies.actor_1_name == "Meryl Streep"]
Meryl_Streep.head()

director_name  num_critic_for_reviews       gross  \
392     Nancy Meyers                   187.0  112.703470   
1038   Curtis Hanson                    42.0   46.815748   
1132     Nora Ephron                   252.0   94.125426   
1322   David Frankel                   208.0  124.732962   
1390  Robert Redford                   227.0   14.998070   

                               genres  actor_1_name             movie_title  \
392              Comedy|Drama|Romance  Meryl Streep       It's Complicated    
1038  Action|Adventure|Crime|Thriller  Meryl Streep         The River Wild    
1132          Biography|Drama|Romance  Meryl Streep          Julie & Julia    
1322             Comedy|Drama|Romance  Meryl Streep  The Devil Wears Prada    
1390               Drama|Thriller|War  Meryl Streep        Lions for Lambs    

      num_voted_users  num_user_for_reviews language  budget  title_year  \
392             69860                 214.0  English    85.0      2009.0   
1038            32544                  69.0  English    45.0      1994.0   
1132            79264                 277.0  English    40.0      2009.0   
1322           286178                 631.0  English    35.0      2006.0   
1390            41170                 298.0  English    35.0      2007.0   

      imdb_score  movie_facebook_likes     profit  
392          6.6                     0  27.703470  
1038         6.3                     0   1.815748  
1132         7.0                 13000  54.125426  
1322         6.8                     0  89.732962  
1390         6.2                     0 -20.001930

In [217]:
# highest imdb score for movies with meryl streep

Meryl_Streep.sort_values('imdb_score', ascending=False)[0:1]

director_name  num_critic_for_reviews     gross         genres  \
1784  Stephen Daldry                   174.0  41.59783  Drama|Romance   

      actor_1_name movie_title  num_voted_users  num_user_for_reviews  \
1784  Meryl Streep  The Hours            102123                 660.0   

     language  budget  title_year  imdb_score  movie_facebook_likes    profit  
1784  English    25.0      2002.0         7.6                     0  16.59783

- The Hours is the highest rated movie with Meryl Streep at 7.6 IMDB score.

In [218]:
#Leo_Caprio 
Leo_Caprio = new_movies[new_movies.actor_1_name == "Leonardo DiCaprio"]
Leo_Caprio.head()

director_name  num_critic_for_reviews       gross  \
25           James Cameron                   315.0  658.672302   
49            Baz Luhrmann                   490.0  144.812796   
94       Christopher Nolan                   642.0  292.568851   
173  Alejandro G. Iñárritu                   556.0  183.635922   
246        Martin Scorsese                   267.0  102.608827   

                               genres       actor_1_name        movie_title  \
25                      Drama|Romance  Leonardo DiCaprio           Titanic    
49                      Drama|Romance  Leonardo DiCaprio  The Great Gatsby    
94   Action|Adventure|Sci-Fi|Thriller  Leonardo DiCaprio         Inception    
173  Adventure|Drama|Thriller|Western  Leonardo DiCaprio      The Revenant    
246                   Biography|Drama  Leonardo DiCaprio       The Aviator    

     num_voted_users  num_user_for_reviews language  budget  title_year  \
25            793059                2528.0  English   200.0      1997.0   
49            362912                 753.0  English   105.0      2013.0   
94           1468200                2803.0  English   160.0      2010.0   
173           406020                1188.0  English   135.0      2015.0   
246           264318                 799.0  English   110.0      2004.0   

     imdb_score  movie_facebook_likes      profit  
25          7.7                 26000  458.672302  
49          7.3                115000   39.812796  
94          8.8                175000  132.568851  
173         8.1                190000   48.635922  
246         7.5                     0   -7.391173

In [219]:
# highest imdb score for movies with leo

Leo_Caprio.sort_values('imdb_score', ascending=False)[0:1]

director_name  num_critic_for_reviews       gross  \
94  Christopher Nolan                   642.0  292.568851   

                              genres       actor_1_name movie_title  \
94  Action|Adventure|Sci-Fi|Thriller  Leonardo DiCaprio  Inception    

    num_voted_users  num_user_for_reviews language  budget  title_year  \
94          1468200                2803.0  English   160.0      2010.0   

    imdb_score  movie_facebook_likes      profit  
94         8.8                175000  132.568851

- Inception is the highest rated movie with Leonardo Dicaprio at 8.8 IMDB score.

In [220]:
#Brad_Pitt
Brad_Pitt = new_movies[new_movies.actor_1_name == "Brad Pitt"]
Brad_Pitt.head()

director_name  num_critic_for_reviews       gross  \
97       David Fincher                   362.0  127.490802   
142  Wolfgang Petersen                   220.0  133.228348   
243  Steven Soderbergh                   198.0  125.531634   
244         Doug Liman                   233.0  186.336103   
367         Tony Scott                   142.0    0.026871   

                                   genres actor_1_name  \
97                  Drama|Fantasy|Romance    Brad Pitt   
142                             Adventure    Brad Pitt   
243                        Crime|Thriller    Brad Pitt   
244  Action|Comedy|Crime|Romance|Thriller    Brad Pitt   
367                 Action|Crime|Thriller    Brad Pitt   

                              movie_title  num_voted_users  \
97   The Curious Case of Benjamin Button            459346   
142                                 Troy            381672   
243                       Ocean's Twelve            284852   
244                     Mr. & Mrs. Smith            348861   
367                             Spy Game            121259   

     num_user_for_reviews language  budget  title_year  imdb_score  \
97                  822.0  English   150.0      2008.0         7.8   
142                1694.0  English   175.0      2004.0         7.2   
243                 627.0  English   110.0      2004.0         6.4   
244                 798.0  English   120.0      2005.0         6.5   
367                 361.0  English    92.0      2001.0         7.0   

     movie_facebook_likes     profit  
97                  23000 -22.509198  
142                     0 -41.771652  
243                     0  15.531634  
244                     0  66.336103  
367                     0 -91.973129

In [221]:
# highest imdb score for movies with brad pitt

Brad_Pitt.sort_values('imdb_score', ascending=False)[0:1]

director_name  num_critic_for_reviews      gross genres actor_1_name  \
646  David Fincher                   315.0  37.023395  Drama    Brad Pitt   

     movie_title  num_voted_users  num_user_for_reviews language  budget  \
646  Fight Club           1347461                2968.0  English    63.0   

     title_year  imdb_score  movie_facebook_likes     profit  
646      1999.0         8.8                 48000 -25.976605

- Highest rated movie with Brad Pitt is Fight Club at 8.8 IMDB score.

In [222]:
# append all three dataframes of the actors

Combined = pd.concat([Meryl_Streep, Leo_Caprio, Brad_Pitt])
Combined

director_name  num_critic_for_reviews       gross  \
392            Nancy Meyers                   187.0  112.703470   
1038          Curtis Hanson                    42.0   46.815748   
1132            Nora Ephron                   252.0   94.125426   
1322          David Frankel                   208.0  124.732962   
1390         Robert Redford                   227.0   14.998070   
1471         Sydney Pollack                    66.0   87.100000   
1514          David Frankel                   234.0   63.536011   
1563          Carl Franklin                    64.0   23.209440   
1784         Stephen Daldry                   174.0   41.597830   
2500         Phyllida Lloyd                   331.0   29.959436   
2793          Robert Altman                   211.0   20.338609   
25            James Cameron                   315.0  658.672302   
49             Baz Luhrmann                   490.0  144.812796   
94        Christopher Nolan                   642.0  292.568851   
173   Alejandro G. Iñárritu                   556.0  183.635922   
246         Martin Scorsese                   267.0  102.608827   
283       Quentin Tarantino                   765.0  162.804648   
293            Edward Zwick                   166.0   57.366262   
294         Martin Scorsese                   606.0  116.866727   
312         Martin Scorsese                   233.0   77.679638   
347         Martin Scorsese                   352.0  132.373442   
431         Martin Scorsese                   490.0  127.968405   
608            Ridley Scott                   238.0   39.380442   
859        Steven Spielberg                   194.0  164.435221   
934             Danny Boyle                   118.0   39.778599   
1046             Sam Mendes                   323.0   22.877808   
1336        Randall Wallace                    83.0   56.876365   
1364         Clint Eastwood                   392.0   37.304950   
1457              Sam Raimi                    63.0   18.636537   
1909             Jerry Zaks                    45.0   12.782508   
2484           Baz Luhrmann                   106.0   46.338728   
3025           Baz Luhrmann                   490.0  144.812796   
97            David Fincher                   362.0  127.490802   
142       Wolfgang Petersen                   220.0  133.228348   
243       Steven Soderbergh                   198.0  125.531634   
244              Doug Liman                   233.0  186.336103   
367              Tony Scott                   142.0    0.026871   
383       Steven Soderbergh                   186.0  183.405771   
448              David Ayer                   406.0   85.707116   
579     Jean-Jacques Annaud                    76.0   37.901509   
646           David Fincher                   315.0   37.023395   
749         Patrick Gilmore                    98.0   26.288320   
887             Neil Jordan                   120.0  105.264608   
1397        Terrence Malick                   584.0   13.303319   
1608         Andrew Dominik                   273.0    3.904982   
2030  Alejandro G. Iñárritu                   285.0   34.300771   
2146    Angelina Jolie Pitt                   131.0    0.531009   
2428         Andrew Dominik                   414.0   14.938570   
2602             Tony Scott                   122.0   12.281500   

                                                 genres       actor_1_name  \
392                                Comedy|Drama|Romance       Meryl Streep   
1038                    Action|Adventure|Crime|Thriller       Meryl Streep   
1132                            Biography|Drama|Romance       Meryl Streep   
1322                               Comedy|Drama|Romance       Meryl Streep   
1390                                 Drama|Thriller|War       Meryl Streep   
1471                            Biography|Drama|Romance       Meryl Streep   
1514                               Comedy|Drama|Romance       Meryl Streep   
1563                                   

In [223]:
# Find mean of critic reviews and audience reviews 

Combined.groupby('actor_1_name')[['num_critic_for_reviews','num_user_for_reviews']].mean(numeric_only=True)

num_critic_for_reviews  num_user_for_reviews
actor_1_name                                                   
Brad Pitt                      245.000000            742.352941
Leonardo DiCaprio              330.190476            914.476190
Meryl Streep                   181.454545            297.181818

- Actor with the highest mean of critic reviews: Leonardo Dicaprio
- Actor with the highest mean of audience reviews: Leonardo Dicaprio

Which actor is highest rated among the three actors according to the user reviews?

In [224]:
# sort descending by user reviews and get top row

Combined.sort_values('num_user_for_reviews', ascending=False)[0:1]

director_name  num_critic_for_reviews      gross genres actor_1_name  \
646  David Fincher                   315.0  37.023395  Drama    Brad Pitt   

     movie_title  num_voted_users  num_user_for_reviews language  budget  \
646  Fight Club           1347461                2968.0  English    63.0   

     title_year  imdb_score  movie_facebook_likes     profit  
646      1999.0         8.8                 48000 -25.976605

The highest rated among the three actors according to the user reviews is Brad Pitt.

Which actor is highest rated among the three actors according to the critics?

In [225]:
# sort descending by critic reviews

Combined.sort_values('num_critic_for_reviews', ascending=False)[0:1]

director_name  num_critic_for_reviews       gross         genres  \
283  Quentin Tarantino                   765.0  162.804648  Drama|Western   

          actor_1_name        movie_title  num_voted_users  \
283  Leonardo DiCaprio  Django Unchained            955174   

     num_user_for_reviews language  budget  title_year  imdb_score  \
283                1193.0  English   100.0      2012.0         8.5   

     movie_facebook_likes     profit  
283                199000  62.804648

The highest rated among the three actors according to the critics is Leonardo Dicaprio.

Which movie starting with the letter A, has the 3rd highest profit?

In [226]:
# using the new_movies dataframe. create new df that includes movies that starts with A
A_movie = new_movies[new_movies['movie_title'].str.startswith('A')]
A_movie.head()

director_name  num_critic_for_reviews       gross  \
0      James Cameron                   723.0  760.505847   
7        Joss Whedon                   635.0  458.991599   
32        Tim Burton                   451.0  334.185206   
59   Robert Zemeckis                   240.0  137.850096   
102      James Bobin                   218.0   76.846624   

                              genres     actor_1_name  \
0    Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
7            Action|Adventure|Sci-Fi  Chris Hemsworth   
32          Adventure|Family|Fantasy      Johnny Depp   
59    Animation|Drama|Family|Fantasy     Robin Wright   
102         Adventure|Family|Fantasy      Johnny Depp   

                          movie_title  num_voted_users  num_user_for_reviews  \
0                             Avatar            886204                3054.0   
7            Avengers: Age of Ultron            462669                1117.0   
32               Alice in Wonderland            306320                 736.0   
59                 A Christmas Carol             72809                 249.0   
102  Alice Through the Looking Glass             21352                 131.0   

    language  budget  title_year  imdb_score  movie_facebook_likes      profit  
0    English   237.0      2009.0         7.9                 33000  523.505847  
7    English   250.0      2015.0         7.5                118000  208.991599  
32   English   200.0      2010.0         6.5                 24000  134.185206  
59   English   200.0      2009.0         6.8                     0  -62.149904  
102  English   170.0      2016.0         6.4                 30000  -93.153376

In [227]:
# sort descending and return movie with the 3rd highest profit
A_movie.sort_values('profit', ascending=False)[2:3]

director_name  num_critic_for_reviews       gross                   genres  \
7   Joss Whedon                   635.0  458.991599  Action|Adventure|Sci-Fi   

      actor_1_name               movie_title  num_voted_users  \
7  Chris Hemsworth  Avengers: Age of Ultron            462669   

   num_user_for_reviews language  budget  title_year  imdb_score  \
7                1117.0  English   250.0      2015.0         7.5   

   movie_facebook_likes      profit  
7                118000  208.991599

The movie starting with the letter A with the 3rd highest profit is Avengers: Age of Ultron.

Among the movies starting with the letter A, who has the highest imdb score?

In [228]:
# sort descending and return the top row

A_movie.sort_values('imdb_score', ascending=False)[0:1]

director_name  num_critic_for_reviews     gross       genres  \
2826     Tony Kaye                   162.0  6.712241  Crime|Drama   

      actor_1_name          movie_title  num_voted_users  \
2826  Ethan Suplee  American History X            782437   

      num_user_for_reviews language  budget  title_year  imdb_score  \
2826                1420.0  English     7.5      1998.0         8.6   

      movie_facebook_likes    profit  
2826                 35000 -0.787759

The movie with the highest imdb score among the movies starting with the letter A is American History X.

Among the movies that '*Joss Whedon*' has directed, which of them made the highest profit?

In [229]:
# get rows where director name is Joss Whedon and sort descending and return the top row

new_movies[new_movies['director_name'] == 'Joss Whedon'].sort_values('profit', ascending=False)[0:1]


director_name  num_critic_for_reviews       gross                   genres  \
16   Joss Whedon                   703.0  623.279547  Action|Adventure|Sci-Fi   

       actor_1_name    movie_title  num_voted_users  num_user_for_reviews  \
16  Chris Hemsworth  The Avengers            995415                1722.0   

   language  budget  title_year  imdb_score  movie_facebook_likes      profit  
16  English   220.0      2012.0         8.1                123000  403.279547

The movie directed by Josh Whedon with the highest profit is The Avengers.

Among the movies that were released in the year '*2007*' & '*2008*', which of them had the 2nd highest imdb score?

In [230]:
# get rows with movies released in 2007 and 2008, sort by imdb_score in descending order and display the 2nd row

new_movies[(new_movies['title_year'] == 2007.0) | (new_movies['title_year'] == 2008.0)].sort_values('imdb_score', ascending=False)[1:2]


director_name  num_critic_for_reviews       gross  \
56  Andrew Stanton                   421.0  223.806889   

                               genres       actor_1_name movie_title  \
56  Adventure|Animation|Family|Sci-Fi  John Ratzenberger     WALL·E    

    num_voted_users  num_user_for_reviews language  budget  title_year  \
56           718837                1043.0  English   180.0      2008.0   

    imdb_score  movie_facebook_likes     profit  
56         8.4                 16000  43.806889

Among the movies that were released in the year '2007' & '2008', the movie with the 2nd highest imdb score is WALL·E.

### Conclusion

- **Action Dominance in Top Profits and IMDb Scores:**
  - Six out of the top 10 highest-profit movies are in the action genre.
  - The movie with the highest profit is "Avatar" by James Cameron.
  - There are two movies directed by James Cameron in the top profit list.
  - The top 5 movies with the highest IMDb scores belong to the crime and action genres.
  - "Star Wars: Episode IV - A New Hope" by George Lucas, ranks 18th but leads in profit, likely due to its large fanbase.
  - The majority of the top 10 highest profits come from action and adventure genres. 
  <br>
  <br>
- **Distribution of Top 200 Movies by IMDb Ratings:**
  - 48 movies have IMDb ratings between 7.5 and 8.
  - 172 movies have IMDb ratings between 8 and 8.5.
  - 44 movies have IMDb ratings between 8.5 and 9.
  - 4 movies have IMDb ratings between 9 and 9.5.
  - No movies have IMDb ratings between 9.5 and 10.
  - The bucket with the maximum number of movies is between 8 and 8.5.
<br>
  <br>
- **Top Rated Movies by Specific Actors:**
  - "The Hours" is the highest-rated movie featuring Meryl Streep, with an IMDb score of 7.6.
  - "Fight Club" is the highest-rated movie featuring Brad Pitt, with an IMDb score of 8.8.
  - "Inception" is the highest-rated movie featuring Leonardo DiCaprio, also with an IMDb score of 8.8.
<br>
  <br>
- **Analysis of Actor Reviews:**
  - Leonardo DiCaprio has the highest mean of both critic and audience reviews.
  - According to user reviews, "Fight Club" featuring Brad Pitt is the highest-rated movie among the three actors.
  - According to critics, "Django Unchained" featuring Leonardo DiCaprio is the highest-rated movie among the three actors.